File: src/main/java/org/janelia/saalfeldlab/paintera/data/n5/N5HDF5Meta.java
package org.janelia.saalfeldlab.paintera.data.n5;

import java.io.IOException;

import ch.systemsx.cisd.hdf5.IHDF5Reader;
import com.google.gson.annotations.Expose;
import org.janelia.saalfeldlab.n5.hdf5.N5HDF5Reader;
import org.janelia.saalfeldlab.n5.hdf5.N5HDF5Writer;

public class N5HDF5Meta implements N5Meta
{

	@Expose
	private final String file;

	@Expose
	private final String dataset;

	@Expose
	private final int[] defaultCellDimensions;

	@Expose
	private final boolean overrideCellDimensions;

	public N5HDF5Meta(final N5HDF5Reader reader, final String dataset) throws ReflectionException
	{
		this(
				ihdfReaderFromReader(reader).getFile().getAbsolutePath().toString(),
				dataset,
				defaultBlockSizeFromReader(reader),
				overrideBlockSizeFromReader(reader)
		    );
	}

	public N5HDF5Meta(final String file, final String dataset, final int[] defaultCellDimensions, final boolean
			overrideCellDimensions)
	{
		super();
		this.file = file;
		this.dataset = dataset;
		this.defaultCellDimensions = defaultCellDimensions;
		this.overrideCellDimensions = overrideCellDimensions;
	}

	@Override
	public N5HDF5Reader reader() throws IOException
	{
		return new N5HDF5Reader(file);
	}

	@Override
	public N5HDF5Writer writer() throws IOException
	{
		return new N5HDF5Writer(file);
	}

	@Override
	public String dataset()
	{
		return dataset;
	}

	private static IHDF5Reader ihdfReaderFromReader(final N5HDF5Reader reader) throws ReflectionException
	{

		try
		{
			return (IHDF5Reader) ReflectionHelpers.searchForField(reader.getClass(), "reader").get(reader);
		} catch (IllegalArgumentException | IllegalAccessException | NoSuchFieldException | SecurityException e)
		{
			throw new ReflectionException(e);
		}
	}

	private static int[] defaultBlockSizeFromReader(final N5HDF5Reader reader) throws ReflectionException
	{

		try
		{
			return (int[]) ReflectionHelpers.searchForField(reader.getClass(), "defaultBlockSize").get(reader);
		} catch (IllegalArgumentException | IllegalAccessException | NoSuchFieldException | SecurityException e)
		{
			throw new ReflectionException(e);
		}
	}

	private static boolean overrideBlockSizeFromReader(final N5HDF5Reader reader) throws ReflectionException
	{

		try
		{
			return (boolean) ReflectionHelpers.searchForField(reader.getClass(), "overrideBlockSize").get(reader);
		} catch (IllegalArgumentException | IllegalAccessException | NoSuchFieldException | SecurityException e)
		{
			throw new ReflectionException(e);
		}
	}

}

File: src/main/kotlin/org/janelia/saalfeldlab/paintera/data/n5/N5HDF5Meta.kt
package org.janelia.saalfeldlab.paintera.data.n5

import com.google.gson.annotations.Expose
import org.janelia.saalfeldlab.n5.hdf5.N5HDF5Reader
import org.janelia.saalfeldlab.n5.hdf5.N5HDF5Writer
import java.io.IOException

class N5HDF5Meta(
    @field:Expose val file: String,
    @field:Expose override val dataset: String,
    @field:Expose private val defaultCellDimensions: IntArray?,
    @field:Expose val isOverrideCellDimensions: Boolean
) : N5Meta {

    val defaultCellDimensionsCopy: IntArray?
        get() = defaultCellDimensions?.clone()

    @Throws(ReflectionException::class)
    constructor(reader: N5HDF5Reader, dataset: String) : this(
        reader.filename.absolutePath,
        dataset,
        reader.defaultBlockSizeCopy,
        reader.doesOverrideBlockSize()
    )

    //FIXME can/should we generify N5Meta rn?
    // TODO this needs to be a reader eventually
    override val reader: N5HDF5Reader
        @Throws(IOException::class)
        get() = writer

    override val writer: N5HDF5Writer
        @Throws(IOException::class)
        get() = defaultCellDimensions?.let { N5HDF5Writer(file, *it) } ?: N5HDF5Writer(file)

    override fun toString() = String.format("{N5HDF5: container=%s dataset=%s}", file, dataset)
}
